app/src/main/java/ch/epfl/skysync/viewmodel/FinishedFlightsViewModel.kt

```diff
@@ -0,0 +1,135 @@
+import ch.epfl.skysync.models.user.User
+import ch.epfl.skysync.util.WhileUiSubscribed
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.launch
+
+/** ViewModel for the user for the finished flights */
+class FinishedFlightsViewModel(val repository: Repository, val userId: String?) : ViewModel() {
+  companion object {
+    @Composable
+    fun createViewModel(
+        repository: Repository,
+        userId: String?,
+    ): FinishedFlightsViewModel {
+      return viewModel<FinishedFlightsViewModel>(
+          factory =
+              object : ViewModelProvider.Factory {
+                override fun <T : ViewModel> create(modelClass: Class<T>): T {
+                  return FinishedFlightsViewModel(repository, userId) as T
+                }
+              })
+    }
+  }
+
+  private val _currentFlights: MutableStateFlow<List<FinishedFlight>?> = MutableStateFlow(null)
+  private val _currentUser: MutableStateFlow<User?> = MutableStateFlow(null)
+  private val _selectedFlightId: MutableStateFlow<String?> = MutableStateFlow(null)
+  private val isLoading = MutableStateFlow(false)
+  private val _flightReports: MutableStateFlow<List<Report>?> = MutableStateFlow(null)
+
+  val currentFlights = _currentFlights.asStateFlow()
+  val currentUser = _currentUser.asStateFlow()
+  val selectedFlightId = _selectedFlightId.asStateFlow()
+  val flightReports = _flightReports.asStateFlow()
+
+  fun refresh() {
+    refreshUserAndFlights()
+  }
+
+  private fun refreshUser() =
+      viewModelScope.launch {
+        _currentUser.value = repository.userTable.get(userId ?: UNSET_ID, onError = { onError(it) })
+      }
+
+  private fun refreshFlights() =
+      viewModelScope.launch {
+        if (_currentUser.value is Admin) {
+          _currentFlights.value =
+              repository.flightTable
+                  .getAll(onError = { onError(it) })
+                  .filterIsInstance<FinishedFlight>()
+        } else if (_currentUser.value is Pilot || _currentUser.value is Crew) {
+          _currentFlights.value =
+              repository.userTable
+                  .retrieveAssignedFlights(
+                      repository.flightTable, userId ?: UNSET_ID, onError = { onError(it) })
+                  .filterIsInstance<FinishedFlight>()
+        }
+      }
+
+  /** Refreshes the user logged in and its finished flights */
+  fun refreshUserAndFlights() =
+      viewModelScope.launch {
+        isLoading.value = true
+        refreshUser().join()
+        refreshFlights().join()
+        isLoading.value = false
+      }
+
+  fun addFlight(flight: FinishedFlight) =
+      viewModelScope.launch {
+        repository.flightTable.add(flight, onError = { onError(it) })
+        refreshUserAndFlights()
+      }
+
+  fun deleteFlight(flightId: String) =
```

Comment:
From what I understand `FinishedFlightsViewModel` is used in the reports screen, when would you need to add/delete a flight there ? These functions already exists in the `FlightViewModel` which is used in the `AdminFlightDetailScreen`